proposals/context.md

```diff
@@ -0,0 +1,265 @@
+# context-protocol
+
+An open protocol for data passing between components.
+
+Author: Benjamin Delarre
+
+Document status: Draft
+
+Last update: 2021-4-27
+
+# Background
+
+There are a number of scenarios where a web component might require data that is provided from outside itself. While components can specify properties and attributes to receive that data imperatively or declaratively, it is often the case that the data may be owned somewhere further up the DOM tree.
+
+One approach to passing this data down to components is commonly referred to as 'prop drilling', whereby components pass properties all the way down through the hierarchy, passing from component to component until it is consumed at its destination. This is generally considered undesirable as it often requires intermediate components in the tree to have knowledge of the data necessary in its descendents.
+
+Frameworks and libraries often provide mechanisms for this, these can range from the simple Context implementation available in React, to more complex Dependency Injection frameworks. Web components need a similar protocol in order to solve this problem.
+
+# Goals
+
+- Allow elements in the DOM to retrieve data based on their contextual position in the DOM
+- Alleviate the problem of 'prop drilling'
+- Simple API that is easily implemented in any framework / library
+- Synchronous protocol, while supporting asynchronous patterns
+- Support single or multiple delivery of context values
+
+# Non-Goals
+
+**Context API !== Dependency Injection Framework**
+
+The Context API does not intend to cover all cases and forms of Dependency Injection. It does not specify constructor, factory or property injection patterns. Its only goal is to formalize the pattern of sharing data across the hierarchy in the DOM, specifically avoiding 'prop drilling' type scenarios. Dependency Injection patterns could be implemented using this protocol, but this is not the goal and should remain explicitly outside the scope of Context API for simplicity.
+
+**Context API is not a state management alternative**
+
+State management libraries often need to perform similar behaviors to the problems that Context API helps to solve. An element deep in the DOM tree made need access to some state, and may need to respond to that state being changed. While state management could be built using the Context API, it is not a primary goal of Context API to solve this problem.
+
+# Details
+
+## The `request-context` event
+
+Components which wish to receive some data from their ancestors should initiate the request by firing a composed, bubbling, `request-context` Event, with a `callback` property.
+
+Typescript interface:
+
+```typescript
+interface RequestContextEvent extends Event {
+  /**
+   * The name of the context that is requested
+   */
+  readonly name: T;
```

Comment:
I think there was some outstanding discussion around string names vs Symbol, yeah? I don't quite remember where that left off. Can we add a todo and an issue? I see some discussion on name conflicts below, but it'd be nice to directly address whether a string key is the best approach and alternatives considered.

Reply:
Yes we haven't really dug into this too far. I've tried to theorize about how things would play out if we used object-values (Symbol, or unique value returned from a 'create' function), but these have all seemed to require some agreed upon central repository of context object keys. This seems somewhat restrictive?

Reply:
I think restrictive is the point. Unless you have some agreement to provide a compatible object, context is going to be too fragile. Importing the same Symbol, which can host docs on the type of object it represents, is a basis for agreement.

I don't think there would need to be any central repository either. All you need is access to the same module. That can happen anywhere, even in a one-liner npm package. That package will serve as the docs for what the context object is supposed to be. If that seems heavyweight, I'd say that even with string keys, if you have cross package collaboration, _something_ needs to describe what that object should be, and that thing can host the symbol.

Reply:
That seems reasonable to me.